src/Build.UnitTests/Evaluation/Evaluator_Tests.cs

```diff
@@ -4328,6 +4328,63 @@ public void VerifyConditionEvaluatorResetStateOnFailure()
                 new BuildEventContext(1, 2, 3, 4)));
         }
 
+        /// <summary>
+        /// Test regression reported at https://github.com/Microsoft/msbuild/issues/2228
+        /// </summary>
+        [Fact]
+        public void ThrownInvalidProjectExceptionProperlyHandled()
+        {
+            string projectContents = ObjectModelHelpers.CleanupFileContents(@"
+                             <Project ToolsVersion=""msbuilddefaulttoolsversion"" xmlns='msbuildnamespace'>
+                                
+                                <Import Project=""import.proj"" />
+
+                            </Project>");
+
+            string importContents = ObjectModelHelpers.CleanupFileContents(@"
+                             <Project ToolsVersion=""msbuilddefaulttoolsversion"" xmlns='msbuildnamespace'>
+
+                                <AnInvalidTopLevelElement />
+
+                              </Project>");
+
+            string projectDirectory = Path.Combine(Path.GetTempPath(), "VerifyTreatAsLocalPropertyUnionBetweenImports");
+
+            try
+            {
+                if (Directory.Exists(projectDirectory))
+                {
+                    FileUtilities.DeleteWithoutTrailingBackslash(projectDirectory, true /* recursive delete */);
+                }
+
+                Directory.CreateDirectory(projectDirectory);
+
+                string primaryProject = Path.Combine(projectDirectory, "project.proj");
+                string import = Path.Combine(projectDirectory, "import.proj");
+
+                File.WriteAllText(primaryProject, projectContents);
+                File.WriteAllText(import, importContents);
+
+                InvalidProjectFileException ex = Assert.Throws<InvalidProjectFileException>( () =>
+                    {
+                        Project unused = new Project(primaryProject, null, null);
+                    })
+                ;
+
+                Assert.Equal($"The element <AnInvalidTopLevelElement> beneath element <Project> is unrecognized.  {import}", ex.Message);
```

Comment:
How about `Contains("<AnInvalidTopLevelElement>")` ? Then we can be confident it's the right error but it's not finding an English string.